File: app/src/main/java/com/pnuema/simplebible/retrievers/BaseRetriever.java
package com.pnuema.simplebible.retrievers;

import android.content.Context;

import java.util.Observable;

public abstract class BaseRetriever extends Observable {
    BaseRetriever() {
    }

    public abstract void getVersions(Context context);

    public abstract void getBooks(Context context);

    public abstract void getChapters(Context context, String book);

    public abstract void getVerses(Context context, String version, String book, String chapter);
}

File: app/src/main/java/com/pnuema/bible/android/retrievers/BaseRetriever.kt
package com.pnuema.bible.android.retrievers

import com.pnuema.bible.android.data.firefly.Books
import com.pnuema.bible.android.data.firefly.ChapterCount
import com.pnuema.bible.android.data.firefly.VerseCount
import com.pnuema.bible.android.data.firefly.Verses
import com.pnuema.bible.android.data.firefly.Versions

abstract class BaseRetriever internal constructor() {

    companion object {
        var instance: BaseRetriever? = null
    }

    val tag: String get() = javaClass.simpleName

    abstract fun savePrefs()

    abstract fun readPrefs()

    abstract suspend fun getVersions(): Versions

    abstract suspend fun getBooks(): Books

    abstract suspend fun getChapters(book: String): ChapterCount

    abstract suspend fun getVerseCount(version: String, book: String, chapter: String): VerseCount

    abstract suspend fun getVerses(version: String, book: String, chapter: String): Verses
}
